File: commands/queue.js
/** @type {import('../').Command} */
export default {
	name: 'queue',
	description: 'Show current queue',
	permissions: {
		memberInVoice: true
	},

	interaction: async ({ interaction, server }) => {
		if (!interaction.guild) return;

		if (!server || server.songs.length < 1) {
			return interaction.reply({
				content: '```nim\nThe queue is empty ;-;\n```',
				ephemeral: true
			});
		}

		return interaction.reply({
			content: buildQueueMsg(server.songs.slice(0, 10)),
			ephemeral: true
		});
	}
};

/** @param {import('../').Song[]} songs */
function buildQueueMsg(songs) {
	let queueMsg = '```nim\n';

	for (let i = 0; i < songs.length; i++) {
		const song = songs[i];
		if (i === 0) {
			queueMsg += '    ⬐ current track\n';
		}

		queueMsg += `${i + 1}) ${
			song.title.length > 40 ? song.title.substring(0, 40 - 1) + '…' : song.title.padEnd(40, ' ')
		} ${song.duration}\n`;
		// TODO show time left for current song `2:39 left`

		if (i === 0) {
			queueMsg += '    ⬑ current track\n';
		}
	}

	queueMsg += '```';
	return queueMsg;
}

File: src/commands/queue.rs
use crate::queue::Queue;
use crate::utils::get_guild_data;
use crate::{check_msg, create_default_message, CommandResult, Context};

/// Display the current song queue.
#[poise::command(slash_command, guild_only)]
pub async fn queue(ctx: Context<'_>) -> CommandResult {
    // TODO: add pagination

    let guild_id = ctx.guild_id().unwrap();

    let guild_data = get_guild_data(ctx, guild_id).await;
    let guild_data = guild_data.lock().await;

    if guild_data.queue.is_empty() {
        let reply = create_default_message("```nim\nThe queue is empty ;-;\n```", true);
        check_msg(ctx.send(reply).await);
        return Ok(());
    }

    let msg = build_queue_msg(&guild_data.queue); // TODO: slice 0..10
    let reply = create_default_message(&msg, true);
    check_msg(ctx.send(reply).await);

    Ok(())
}

fn build_queue_msg(queue: &Queue) -> String {
    let mut msg = String::from("```nim\n");
    let max = queue.len().min(10);

    for (i, song) in queue.tracks.iter().take(max).enumerate() {
        if i == 0 {
            msg.push_str("    ⬐ current track\n");
        }

        let mut title = song.title.clone();
        if title.chars().count() > 40 {
            title = title.chars().take(39).collect::<String>() + "…";
        } else {
            title = format!("{:width$}", title, width = 40);
        }

        // TOOD: show time left

        msg.push_str(&format!("{}) {} {}\n", i + 1, title, song.duration));

        if i == 0 {
            msg.push_str("    ⬑ current track\n");
        }
    }

    msg.push_str("```");
    msg
}
